pkg/orchestrator/assetscanestimationwatcher/watcher.go

```diff
@@ -0,0 +1,269 @@
+	AssetScanEstimationReconciler = common.Reconciler[AssetScanEstimationReconcileEvent]
+)
+
+func New(c Config) *Watcher {
+	return &Watcher{
+		backend:          c.Backend,
+		provider:         c.Provider,
+		pollPeriod:       c.PollPeriod,
+		reconcileTimeout: c.ReconcileTimeout,
+		abortTimeout:     c.AbortTimeout,
+		queue:            common.NewQueue[AssetScanEstimationReconcileEvent](),
+	}
+}
+
+type Watcher struct {
+	backend          *backendclient.BackendClient
+	provider         provider.Provider
+	pollPeriod       time.Duration
+	reconcileTimeout time.Duration
+	abortTimeout     time.Duration
+
+	queue *AssetScanEstimationQueue
+}
+
+func (w *Watcher) Start(ctx context.Context) {
+	logger := log.GetLoggerFromContextOrDiscard(ctx).WithField("controller", "AssetScanEstimationWatcher")
+	ctx = log.SetLoggerForContext(ctx, logger)
+
+	poller := &AssetScanEstimationPoller{
+		PollPeriod: w.pollPeriod,
+		Queue:      w.queue,
+		GetItems:   w.GetAssetScanEstimations,
+	}
+	poller.Start(ctx)
+
+	reconciler := &AssetScanEstimationReconciler{
+		ReconcileTimeout:  w.reconcileTimeout,
+		Queue:             w.queue,
+		ReconcileFunction: w.Reconcile,
+	}
+	reconciler.Start(ctx)
+}
+
+// nolint:cyclop
+func (w *Watcher) GetAssetScanEstimations(ctx context.Context) ([]AssetScanEstimationReconcileEvent, error) {
+	logger := log.GetLoggerFromContextOrDiscard(ctx)
+	logger.Debugf("Fetching AssetScanEstimations which need to be reconciled")
+
+	filter := fmt.Sprintf("state/state ne '%s' and state/state ne '%s'",
+		models.AssetScanEstimationStateStateDone, models.AssetScanEstimationStateStateFailed)
+	selector := "id,scanEstimation/id,asset/id"
+	params := models.GetAssetScanEstimationsParams{
+		Filter: &filter,
+		Select: &selector,
+		Count:  utils.PointerTo(true),
+	}
+	assetScanEstimations, err := w.backend.GetAssetScanEstimations(ctx, params)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get AssetScanEstimations: %w", err)
+	}
+
+	switch {
+	case assetScanEstimations.Items == nil && assetScanEstimations.Count == nil:
+		return nil, fmt.Errorf("failed to fetch AssetScanEstimations: invalid API response: %v", assetScanEstimations)
+	case assetScanEstimations.Count != nil && *assetScanEstimations.Count <= 0:
+		fallthrough
+	case assetScanEstimations.Items != nil && len(*assetScanEstimations.Items) <= 0:
+		return nil, nil
+	}
+
+	events := make([]AssetScanEstimationReconcileEvent, 0, len(*assetScanEstimations.Items))
+	for _, assetScanEstimation := range *assetScanEstimations.Items {
+		assetScanEstimationID, ok := assetScanEstimation.GetID()
+		if !ok {
+			logger.Warnf("Skipping due to invalid AssetScanEstimation: ID is nil: %v", assetScanEstimation)
+			continue
+		}
+		scanEstimationID, ok := assetScanEstimation.GetScanEstimationID()
+		if !ok {
+			logger.Warnf("Skipping due to invalid AssetScanEstimation: ScanEstimation.ID is nil: %v", assetScanEstimation)
+			continue
+		}
```

Comment:
AssetScanEstimations should be usable by themselves to estimate a single asset, ScanEstimationID is/should be optional to support this use case and no where in the asset scan estimation watcher should use it outside of informational logging.

Reply:
Took that logic from AssetScan watcher. Should we fix this in there as well? (We have single assetScans as part of the CLI flow, no?)

Reply:
It should be removed there too, I think that was missed as part of the "use AssetScans independently" work, I'll open a ticket for it.

Reply:
https://github.com/openclarity/vmclarity/issues/607

Reply:
It is an Info log, just for informational purposes